aten/src/ATen/core/dispatch/OpSchemaRegistration.h

```diff
@@ -2,17 +2,38 @@
 
 #include <ATen/core/dispatch/Dispatcher.h>
 
-// TODO Better error message when this definition is missing
+namespace c10 {
+namespace detail {
+class OpSchemaRegistrar final {
+public:
+  explicit OpSchemaRegistrar(FunctionSchema schema)
+  : opHandle_(c10::Dispatcher::singleton().registerSchema(std::move(schema))) {}
+
+  ~OpSchemaRegistrar() {
+    c10::Dispatcher::singleton().deregisterSchema(opHandle_);
+  }
+
+  const c10::OperatorHandle& opHandle() const {
+    return opHandle_;
+  }
+
+private:
+  c10::OperatorHandle opHandle_;
+};
+}
+}
 
 /**
- * Macro for defining an operator schema.  Every user-defined OpSchemaDef struct must
- * invoke this macro on it.  Internally, this arranges for the dispatch table for
+ * Macro for defining an operator schema.  Every operator schema must
+ * invoke C10_DECLARE_OP_SCHEMA in a header and C10_DEFINE_OP_SCHEMA in one (!)
+ * cpp file.  Internally, this arranges for the dispatch table for
  * the operator to be created.
  */
-#define C10_DEFINE_OP_SCHEMA(OpSchemaDef)                                         \
-  template<>                                                                      \
-  C10_EXPORT c10::DispatchTable<OpSchemaDef>& c10_dispatch_table<OpSchemaDef>() { \
-    static c10::DispatchTable<OpSchemaDef> singleton;                             \
-    return singleton;                                                             \
+#define C10_DECLARE_OP_SCHEMA(Name)                                             \
+  CAFFE2_API const c10::OperatorHandle& Name();                                 \
```

Comment:
Shouldn't this be `C10_IMPORT`? At least, that's how it's declared in `c10/core/Registry.h`

Reply:
That's weird. cc @Yangqing . Afaik, this has to be `_API` because the declaration in the header, when included from inside the library, should be `_EXPORT` not `_IMPORT`. Should only be `_IMPORT` when included from outside the library.